README.md
sabre/vobject
=============

The VObject library allows you to easily parse and manipulate [iCalendar](https://tools.ietf.org/html/rfc5545)
and [vCard](https://tools.ietf.org/html/rfc6350) objects using PHP.

The goal of the VObject library is to create a very complete library, with an easy to use API.

Derivative
----------

This derivative of Sabre\vObject parses data from legacy libical versions correctly.


Installation
------------

VObject requires PHP 5.3, and should be installed using composer.
The general composer instructions can be found on the [composer website](http://getcomposer.org/doc/00-intro.md composer website).

After that, just declare the vobject dependency as follows:

    "require" : {
        "sabre/vobject" : "~3.4"
    }

Then, run `composer.phar update` and you should be good.

Usage
-----

* [3.x documentation](http://sabre.io/vobject/usage/)
* [2.x documentation](http://sabre.io/vobject/usage_2/)
* [Migrating from 2.x to 3.x](http://sabre.io/vobject/upgrade/)

Support
-------

Head over to the [SabreDAV mailing list](http://groups.google.com/group/sabredav-discuss) for any questions.

Made at fruux
-------------

This library is being developed by [fruux](https://fruux.com/). Drop us a line for commercial services or enterprise support.


ChangeLog.md
ChangeLog
=========

3.5.3+SimKolab (2016-12-15)
---------------------------

* #134: Really check for all C0 control characters.
* Fix UTF-8 validator.
* Convert using cp1252 as fallback encoding.
* #129: Parse libical-specific or otherwise invalid escapes in TEXT the
  same way libical does.

3.5.3 (2016-10-06)
------------------

* #331: Fix dealing with multiple overridden instances falling on the same
  date/time (@afedyk-sugarcrm).


3.5.2 (2016-04-24)
-----------------

* #312: Backported a fix related to iTip processing of events with timezones,
  without a master event.


3.5.1 (2016-04-06)
------------------

* #309: When expanding recurring events, the first event should also have a
  `RECURRENCE-ID` property.
* #306: iTip REPLYs to the first instance of a recurring event was not handled
  correctly.


3.5.0 (2016-01-11)
------------------

* This release supports PHP 7, contrary to 3.4.x versions.
* BC Break: `Sabre\VObject\Property\Float` has been renamed to
  `Sabre\VObject\Property\FloatValue`.
* BC Break: `Sabre\VObject\Property\Integer` has been renamed to
  `Sabre\VObject\Property\IntegerValue`.


3.4.9 (2016-01-11)
------------------

* This package now specifies in composer.json that it does not support PHP 7.
  For PHP 7, use version 3.5.x or 4.x.


3.4.8 (2016-01-04)
------------------

* #284: When generating `CANCEL` iTip messages, we now include `DTEND`.
  (@kewisch).


3.4.7 (2015-09-05)
------------------

* #253: Handle `isInTimeRange` for recurring events that have 0 valid
  instances. (@DominikTo, @migrax).


3.4.6 (2015-08-06)
------------------

* #250: Recurring all-day events are incorrectly included in time range
  requests when not using UTC in the time range. (@armin-hackmann)


3.4.5 (2015-06-02)
------------------

* #229: Converting vcards from 3.0 to 4.0 that contained a `LANG` property
  would throw an error.


3.4.4 (2015-05-27)
------------------

* #228: Fixed a 'party crasher' bug in the iTip broker. This would break
  scheduling in some cases.


3.4.3 (2015-05-19)
------------------

* #219: Corrected validation of `EXDATE` properties with more than one value.
* #212: `BYSETPOS` with values below `-1` was broken and could cause infinite
  loops.
* #211: Fix `BYDAY=-5TH` in recurrence iterator. (@lindquist)
* #216: `ENCODING` parameter is now validated for all document types.
* #217: Initializing vCard `DATE` objects with a PHP DateTime object will now
  work correctly. (@thomascube)


3.4.2 (2015-02-25)
------------------

* #210: iTip: Replying to an event without a master event was broken.


3.4.1 (2015-02-24)
------------------

* A minor change to ensure that unittests work correctly in the sabre/dav
  test-suite.


3.4.0 (2015-02-23)
------------------

* #196: Made parsing recurrence rules a lot faster on big calendars.
* Updated windows timezone mappings to latest unicode version.
* #202: Support for parsing and validating `VAVAILABILITY` components. (@Hywan)
* #195: PHP 5.3 compatibility in 'generatevcards' script. (@rickdenhaan)
* #205: Improving handling of multiple `EXDATE` when processing iTip changes.
  (@armin-hackmann)
* #187: Fixed validator rules for `LAST-MODIFIED` properties.
* #188: Retain floating times when generating instances using
  `Recur\EventIterator`.
* #203: Skip tests for timezones that are not supported on older PHP versions,
  instead of a hard fail.
* #204: Dealing a bit better with vCard date-time values that contained
  milliseconds. (which is normally invalid). (@armin-hackmann)


3.3.5 (2015-01-09)
------------------

* #168: Expanding calendars now removes objects with recurrence rules that
  don't have a valid recurrence instance.
* #177: SCHEDULE-STATUS should not contain a reason phrase, only a status
  code.
* #175: Parser can now read and skip the UTF-8 BOM.
* #179: Added `isFloating` to `DATE-TIME` properties.
* #179: Fixed jCal serialization of floating `DATE-TIME` properties.
* #173: vCard converter failed for `X-ABDATE` properties that had no
  `X-ABLABEL`.
* #180: Added `PROFILE_CALDAV` and `PROFILE_CARDDAV` to enable validation rules
  specific for CalDAV/CardDAV servers.
* #176: A missing `UID` is no longer an error, but a warning for the vCard
  validator, unless `PROFILE_CARDDAV` is specified.


3.3.4 (2014-11-19)
------------------

* #154: Converting `ANNIVERSARY` to `X-ANNIVERSARY` and `X-ABDATE` and
  vice-versa when converting to/from vCard 4.
* #154: It's now possible to easily select all vCard properties belonging to
  a single group with `$vcard->{'ITEM1.'}` syntax. (@armin-hackmann)
* #156: Simpler way to check if a string is UTF-8. (@Hywan)
* Unittest improvements.
* #159: The recurrence iterator, freebusy generator and iCalendar DATE and

tests/VObject/Parser/InvalidEncodingTest.php
<?php

namespace Sabre\VObject\Parser;

use
    Sabre\VObject\Reader;

class InvalidEncodingTest extends \PHPUnit_Framework_TestCase {

    function testLibicalSpecificValues() {

        $data = "BEGIN:VCARD\nLABEL:a'b\"c\\\\d\\;e\\,f:g^h&i<j>k!l\\\"M\\tN\\rO\\bP\\fQ\nEND:VCARD";

        /* with compatibility kludges */
        $result = Reader::read($data);
        $this->assertEquals("a'b\"c\\d;e,f:g^h&i<j>k!l\"M\tN\rO\010P\fQ", "".$result->LABEL);

    }
}


lib/timezonedata/windowszones.php
<?php

/**
 * Automatically generated timezone file
 *
 * Last update: 2016-08-24T17:35:38-04:00
 * Source: http://unicode.org/repos/cldr/trunk/common/supplemental/windowsZones.xml
 *
 * @copyright Copyright (C) 2011-2015 fruux GmbH (https://fruux.com/).
 * @license http://sabre.io/license/ Modified BSD License
 */

return array(
  'AUS Central Standard Time'       => 'Australia/Darwin',
  'AUS Eastern Standard Time'       => 'Australia/Sydney',
  'Afghanistan Standard Time'       => 'Asia/Kabul',
  'Alaskan Standard Time'           => 'America/Anchorage',
  'Aleutian Standard Time'          => 'America/Adak',
  'Altai Standard Time'             => 'Asia/Barnaul',
  'Arab Standard Time'              => 'Asia/Riyadh',
  'Arabian Standard Time'           => 'Asia/Dubai',
  'Arabic Standard Time'            => 'Asia/Baghdad',
  'Argentina Standard Time'         => 'America/Buenos_Aires',
  'Astrakhan Standard Time'         => 'Europe/Astrakhan',
  'Atlantic Standard Time'          => 'America/Halifax',
  'Aus Central W. Standard Time'    => 'Australia/Eucla',
  'Azerbaijan Standard Time'        => 'Asia/Baku',
  'Azores Standard Time'            => 'Atlantic/Azores',
  'Bahia Standard Time'             => 'America/Bahia',
  'Bangladesh Standard Time'        => 'Asia/Dhaka',
  'Belarus Standard Time'           => 'Europe/Minsk',
  'Bougainville Standard Time'      => 'Pacific/Bougainville',
  'Canada Central Standard Time'    => 'America/Regina',
  'Cape Verde Standard Time'        => 'Atlantic/Cape_Verde',
  'Caucasus Standard Time'          => 'Asia/Yerevan',
  'Cen. Australia Standard Time'    => 'Australia/Adelaide',
  'Central America Standard Time'   => 'America/Guatemala',
  'Central Asia Standard Time'      => 'Asia/Almaty',
  'Central Brazilian Standard Time' => 'America/Cuiaba',
  'Central Europe Standard Time'    => 'Europe/Budapest',
  'Central European Standard Time'  => 'Europe/Warsaw',
  'Central Pacific Standard Time'   => 'Pacific/Guadalcanal',
  'Central Standard Time'           => 'America/Chicago',
  'Central Standard Time (Mexico)'  => 'America/Mexico_City',
  'Chatham Islands Standard Time'   => 'Pacific/Chatham',
  'China Standard Time'             => 'Asia/Shanghai',
  'Cuba Standard Time'              => 'America/Havana',
  'Dateline Standard Time'          => 'Etc/GMT+12',
  'E. Africa Standard Time'         => 'Africa/Nairobi',
  'E. Australia Standard Time'      => 'Australia/Brisbane',
  'E. Europe Standard Time'         => 'Europe/Chisinau',
  'E. South America Standard Time'  => 'America/Sao_Paulo',
  'Easter Island Standard Time'     => 'Pacific/Easter',
  'Eastern Standard Time'           => 'America/New_York',
  'Eastern Standard Time (Mexico)'  => 'America/Cancun',
  'Egypt Standard Time'             => 'Africa/Cairo',
  'Ekaterinburg Standard Time'      => 'Asia/Yekaterinburg',
  'FLE Standard Time'               => 'Europe/Kiev',
  'Fiji Standard Time'              => 'Pacific/Fiji',
  'GMT Standard Time'               => 'Europe/London',
  'GTB Standard Time'               => 'Europe/Bucharest',
  'Georgian Standard Time'          => 'Asia/Tbilisi',
  'Greenland Standard Time'         => 'America/Godthab',
  'Greenwich Standard Time'         => 'Atlantic/Reykjavik',
  'Haiti Standard Time'             => 'America/Port-au-Prince',
  'Hawaiian Standard Time'          => 'Pacific/Honolulu',
  'India Standard Time'             => 'Asia/Calcutta',
  'Iran Standard Time'              => 'Asia/Tehran',
  'Israel Standard Time'            => 'Asia/Jerusalem',
  'Jordan Standard Time'            => 'Asia/Amman',
  'Kaliningrad Standard Time'       => 'Europe/Kaliningrad',
  'Korea Standard Time'             => 'Asia/Seoul',
  'Libya Standard Time'             => 'Africa/Tripoli',
  'Line Islands Standard Time'      => 'Pacific/Kiritimati',
  'Lord Howe Standard Time'         => 'Australia/Lord_Howe',
  'Magadan Standard Time'           => 'Asia/Magadan',
  'Marquesas Standard Time'         => 'Pacific/Marquesas',
  'Mauritius Standard Time'         => 'Indian/Mauritius',
  'Middle East Standard Time'       => 'Asia/Beirut',
  'Montevideo Standard Time'        => 'America/Montevideo',
  'Morocco Standard Time'           => 'Africa/Casablanca',
  'Mountain Standard Time'          => 'America/Denver',
  'Mountain Standard Time (Mexico)' => 'America/Chihuahua',
  'Myanmar Standard Time'           => 'Asia/Rangoon',
  'N. Central Asia Standard Time'   => 'Asia/Novosibirsk',
  'Namibia Standard Time'           => 'Africa/Windhoek',
  'Nepal Standard Time'             => 'Asia/Katmandu',
  'New Zealand Standard Time'       => 'Pacific/Auckland',
  'Newfoundland Standard Time'      => 'America/St_Johns',
  'Norfolk Standard Time'           => 'Pacific/Norfolk',
  'North Asia East Standard Time'   => 'Asia/Irkutsk',
  'North Asia Standard Time'        => 'Asia/Krasnoyarsk',
  'North Korea Standard Time'       => 'Asia/Pyongyang',
  'Pacific SA Standard Time'        => 'America/Santiago',
  'Pacific Standard Time'           => 'America/Los_Angeles',
  'Pacific Standard Time (Mexico)'  => 'America/Tijuana',
  'Pakistan Standard Time'          => 'Asia/Karachi',
  'Paraguay Standard Time'          => 'America/Asuncion',
  'Romance Standard Time'           => 'Europe/Paris',
  'Russia Time Zone 10'             => 'Asia/Srednekolymsk',
  'Russia Time Zone 11'             => 'Asia/Kamchatka',
  'Russia Time Zone 3'              => 'Europe/Samara',
  'Russian Standard Time'           => 'Europe/Moscow',
  'SA Eastern Standard Time'        => 'America/Cayenne',
  'SA Pacific Standard Time'        => 'America/Bogota',
  'SA Western Standard Time'        => 'America/La_Paz',
  'SE Asia Standard Time'           => 'Asia/Bangkok',
  'Saint Pierre Standard Time'      => 'America/Miquelon',
  'Sakhalin Standard Time'          => 'Asia/Sakhalin',
  'Samoa Standard Time'             => 'Pacific/Apia',
  'Singapore Standard Time'         => 'Asia/Singapore',
  'South Africa Standard Time'      => 'Africa/Johannesburg',
  'Sri Lanka Standard Time'         => 'Asia/Colombo',
  'Syria Standard Time'             => 'Asia/Damascus',
  'Taipei Standard Time'            => 'Asia/Taipei',
  'Tasmania Standard Time'          => 'Australia/Hobart',
  'Tocantins Standard Time'         => 'America/Araguaina',
  'Tokyo Standard Time'             => 'Asia/Tokyo',
  'Tomsk Standard Time'             => 'Asia/Tomsk',
  'Tonga Standard Time'             => 'Pacific/Tongatapu',
  'Transbaikal Standard Time'       => 'Asia/Chita',
  'Turkey Standard Time'            => 'Europe/Istanbul',
  'Turks And Caicos Standard Time'  => 'America/Grand_Turk',
  'US Eastern Standard Time'        => 'America/Indianapolis',
  'US Mountain Standard Time'       => 'America/Phoenix',
  'UTC'                             => 'Etc/GMT',
  'UTC+12'                          => 'Etc/GMT-12',
  'UTC-02'                          => 'Etc/GMT+2',
  'UTC-08'                          => 'Etc/GMT+8',
  'UTC-09'                          => 'Etc/GMT+9',
  'UTC-11'                          => 'Etc/GMT+11',
  'Ulaanbaatar Standard Time'       => 'Asia/Ulaanbaatar',
  'Venezuela Standard Time'         => 'America/Caracas',
  'Vladivostok Standard Time'       => 'Asia/Vladivostok',
  'W. Australia Standard Time'      => 'Australia/Perth',
  'W. Central Africa Standard Time' => 'Africa/Lagos',
  'W. Europe Standard Time'         => 'Europe/Berlin',
  'W. Mongolia Standard Time'       => 'Asia/Hovd',
  'West Asia Standard Time'         => 'Asia/Tashkent',
  'West Bank Standard Time'         => 'Asia/Hebron',
  'West Pacific Standard Time'      => 'Pacific/Port_Moresby',
  'Yakutsk Standard Time'           => 'Asia/Yakutsk',
);


lib/Parser/MimeDir.php
<?php

namespace Sabre\VObject\Parser;

use
    Sabre\VObject\ParseException,
    Sabre\VObject\EofException,
    Sabre\VObject\Component,
    Sabre\VObject\Property,
    Sabre\VObject\Component\VCalendar,
    Sabre\VObject\Component\VCard;

/**
 * MimeDir parser.
 *
 * This class parses iCalendar 2.0 and vCard 2.1, 3.0 and 4.0 files. This
 * parser will return one of the following two objects from the parse method:
 *
 * Sabre\VObject\Component\VCalendar
 * Sabre\VObject\Component\VCard
 *
 * Copyright (C) 2011-2016 Evert Pot (http://evertpot.com/),
 *	fruux GmbH (https://fruux.com/)
 * Copyright © 2014, 2016 mirabilos,
 *	tarent solutions GmbH (https://www.tarent.de/)
 *
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 * - Redistributions of source code must retain the above copyright notice,
 *   this list of conditions and the following disclaimer.
 * - Redistributions in binary form must reproduce the above copyright notice,
 *   this list of conditions and the following disclaimer in the documentation
 *   and/or other materials provided with the distribution.
 * - Neither the name Sabre nor the names of its contributors
 *   may be used to endorse or promote products derived from this software
 *   without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
 * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */
class MimeDir extends Parser {

    /**
     * The input stream.
     *
     * @var resource
     */
    protected $input;

    /**
     * Root component
     *
     * @var Component
     */
    protected $root;

    /**
     * Parses an iCalendar or vCard file
     *
     * Pass a stream or a string. If null is parsed, the existing buffer is
     * used.
     *
     * @param string|resource|null $input
     * @param int|null $options
     * @return array
     */
    public function parse($input = null, $options = null) {

        $this->root = null;
        if (!is_null($input)) {

            $this->setInput($input);

        }

        if (!is_null($options)) $this->options = $options;

        $this->parseDocument();

        return $this->root;

    }

    /**
     * Sets the input buffer. Must be a string or stream.
     *
     * @param resource|string $input
     * @return void
     */
    public function setInput($input) {

        // Resetting the parser
        $this->lineIndex = 0;
        $this->startLine = 0;

        if (is_string($input)) {
            // Convering to a stream.
            $stream = fopen('php://temp', 'r+');
            fwrite($stream, $input);
            rewind($stream);
            $this->input = $stream;
        } elseif (is_resource($input)) {
            $this->input = $input;
        } else {
            throw new \InvalidArgumentException('This parser can only read from strings or streams.');
        }

    }

    /**
     * Parses an entire document.
     *
     * @return void
     */
    protected function parseDocument() {

        $line = $this->readLine();

        // BOM is ZERO WIDTH NO-BREAK SPACE (U+FEFF).
        // It's 0xEF 0xBB 0xBF in UTF-8 hex.
        if (   3 <= strlen($line)
            && ord($line[0]) === 0xef
            && ord($line[1]) === 0xbb
            && ord($line[2]) === 0xbf) {
            $line = substr($line, 3);
        }

        switch(strtoupper($line)) {
            case 'BEGIN:VCALENDAR' :
                $class = isset(VCalendar::$componentMap['VCALENDAR'])
                    ? VCalendar::$componentMap[$name]
                    : 'Sabre\\VObject\\Component\\VCalendar';
                break;
            case 'BEGIN:VCARD' :
                $class = isset(VCard::$componentMap['VCARD'])
                    ? VCard::$componentMap['VCARD']
                    : 'Sabre\\VObject\\Component\\VCard';
                break;
            default :
                throw new ParseException('This parser only supports VCARD and VCALENDAR files');
        }

        $this->root = new $class(array(), false);

        while(true) {

            // Reading until we hit END:
            $line = $this->readLine();
            if (strtoupper(substr($line,0,4)) === 'END:') {
                break;
            }
            $result = $this->parseLine($line);
            if ($result) {
                $this->root->add($result);
            }

        }

        $name = strtoupper(substr($line, 4));
        if ($name!==$this->root->name) {
            throw new ParseException('Invalid MimeDir file. expected: "END:' . $this->root->name . '" got: "END:' . $name . '"');
        }

    }

    /**
     * Parses a line, and if it hits a component, it will also attempt to parse
     * the entire component
     *
     * @param string $line Unfolded line
     * @return Node
     */
    protected function parseLine($line) {

        // Start of a new component
        if (strtoupper(substr($line, 0, 6)) === 'BEGIN:') {

            $component = $this->root->createComponent(substr($line,6), array(), false);

            while(true) {

                // Reading until we hit END:
                $line = $this->readLine();
                if (strtoupper(substr($line,0,4)) === 'END:') {
                    break;
                }
                $result = $this->parseLine($line);
                if ($result) {
                    $component->add($result);
                }

            }

            $name = strtoupper(substr($line, 4));
            if ($name!==$component->name) {
                throw new ParseException('Invalid MimeDir file. expected: "END:' . $component->name . '" got: "END:' . $name . '"');
            }

            return $component;

        } else {

            // Property reader
            $property = $this->readProperty($line);
            if (!$property) {
                // Ignored line
                return false;
            }
            return $property;

        }

    }

    /**
     * We need to look ahead 1 line every time to see if we need to 'unfold'
     * the next line.
     *
     * If that was not the case, we store it here.
     *
     * @var null|string
     */
    protected $lineBuffer;

    /**
     * The real current line number.
     */
    protected $lineIndex = 0;

    /**
     * In the case of unfolded lines, this property holds the line number for
     * the start of the line.
     *
     * @var int
     */
    protected $startLine = 0;

    /**
     * Contains a 'raw' representation of the current line.
     *
     * @var string
     */
    protected $rawLine;

    /**
     * Reads a single line from the buffer.
     *
     * This method strips any newlines and also takes care of unfolding.
     *
     * @throws \Sabre\VObject\EofException
     * @return string
     */
    protected function readLine() {

        if (!is_null($this->lineBuffer)) {
            $rawLine = $this->lineBuffer;
            $this->lineBuffer = null;
        } else {
            do {
                $eof = feof($this->input);

                $rawLine = fgets($this->input);

                if ($eof || (feof($this->input) && $rawLine===false)) {
                    throw new EofException('End of document reached prematurely');
                }
                if ($rawLine === false) {
                    throw new ParseException('Error reading from input stream');
                }
                $rawLine = rtrim($rawLine, "\r\n");
            } while ($rawLine === ''); // Skipping empty lines
            $this->lineIndex++;
        }
        $line = $rawLine;

        $this->startLine = $this->lineIndex;

        // Looking ahead for folded lines.
        while (true) {

            $nextLine = rtrim(fgets($this->input), "\r\n");
            $this->lineIndex++;
            if (!$nextLine) {
                break;
            }
            if ($nextLine[0] === "\t" || $nextLine[0] === " ") {
                $line .= substr($nextLine, 1);
                $rawLine .= "\n " . substr($nextLine, 1);
            } else {
                $this->lineBuffer = $nextLine;
                break;
            }

        }
        $this->rawLine = $rawLine;
        return $line;

    }

    /**
     * Reads a property or component from a line.
     *
     * @return void
     */
    protected function readProperty($line) {

        if ($this->options & self::OPTION_FORGIVING) {
            $propNameToken = 'A-Z0-9\-\._\\/';
        } else {
            $propNameToken = 'A-Z0-9\-\.';
        }

        $paramNameToken = 'A-Z0-9\-';
        $safeChar = '^";:,';
        $qSafeChar = '^"';

        $regex = "/
            ^(?P<name> [$propNameToken]+ ) (?=[;:])        # property name
            |
            (?<=:)(?P<propValue> .+)$                      # property value
            |
            ;(?P<paramName> [$paramNameToken]+) (?=[=;:])  # parameter name
            |
            (=|,)(?P<paramValue>                           # parameter value
                (?: [$safeChar]*) |
                \"(?: [$qSafeChar]+)\"
            ) (?=[;:,])
            /xi";

        //echo $regex, "\n"; die();
        preg_match_all($regex, $line, $matches,  PREG_SET_ORDER);

        $property = array(
            'name' => null,
            'parameters' => array(),
            'value' => null
        );

        $lastParam = null;

        /**
         * Looping through all the tokens.
         *
         * Note that we are looping through them in reverse order, because if a
         * sub-pattern matched, the subsequent named patterns will not show up
         * in the result.
         */
        foreach($matches as $match) {

            if (isset($match['paramValue'])) {
                if ($match['paramValue'] && $match['paramValue'][0] === '"') {
                    $value = substr($match['paramValue'], 1, -1);
                } else {
                    $value = $match['paramValue'];
                }

                $value = $this->unescapeParam($value);

                if (is_null($property['parameters'][$lastParam])) {
                    $property['parameters'][$lastParam] = $value;
                } elseif (is_array($property['parameters'][$lastParam])) {
                    $property['parameters'][$lastParam][] = $value;
                } else {
                    $property['parameters'][$lastParam] = array(
                        $property['parameters'][$lastParam],
                        $value
                    );
                }
                continue;
            }
            if (isset($match['paramName'])) {
                $lastParam = strtoupper($match['paramName']);
                if (!isset($property['parameters'][$lastParam])) {
                    $property['parameters'][$lastParam] = null;
                }
                continue;
            }
            if (isset($match['propValue'])) {
                $property['value'] = $match['propValue'];
                continue;
            }
            if (isset($match['name']) && $match['name']) {
                $property['name'] = strtoupper($match['name']);
                continue;
            }

            // @codeCoverageIgnoreStart
            throw new \LogicException('This code should not be reachable');
            // @codeCoverageIgnoreEnd

        }

        if (is_null($property['value'])) {
            $property['value'] = '';
        }
        if (!$property['name']) {
            if ($this->options & self::OPTION_IGNORE_INVALID_LINES) {
                return false;
            }
            throw new ParseException('Invalid Mimedir file. Line starting at ' . $this->startLine . ' did not follow iCalendar/vCard conventions');
        }

        // vCard 2.1 states that parameters may appear without a name, and only
        // a value. We can deduce the value based on it's name.
        //
        // Our parser will get those as parameters without a value instead, so
        // we're filtering these parameters out first.
        $namedParameters = array();
        $namelessParameters = array();

        foreach($property['parameters'] as $name=>$value) {
            if (!is_null($value)) {
                $namedParameters[$name] = $value;
            } else {
                $namelessParameters[] = $name;
            }
        }

        $propObj = $this->root->createProperty($property['name'], null, $namedParameters);

        foreach($namelessParameters as $namelessParameter) {
            $propObj->add(null, $namelessParameter);
        }

        if (strtoupper($propObj['ENCODING']) === 'QUOTED-PRINTABLE') {
            $propObj->setQuotedPrintableValue($this->extractQuotedPrintableValue());
        } else {
            $propObj->setRawMimeDirValue($property['value']);
        }

        return $propObj;

    }

    /**
     * Unescapes a property value.
     *
     * vCard 2.1 says:
     *   * Semi-colons must be escaped in some property values, specifically
     *     ADR, ORG and N.
     *   * Semi-colons must be escaped in parameter values, because semi-colons
     *     are also use to separate values.
     *   * No mention of escaping backslashes with another backslash.
     *   * newlines are not escaped either, instead QUOTED-PRINTABLE is used to
     *     span values over more than 1 line.
     *
     * vCard 3.0 says:
     *   * (rfc2425) Backslashes, newlines (\n or \N) and comma's must be
     *     escaped, all time time.
     *   * Comma's are used for delimeters in multiple values
     *   * (rfc2426) Adds to to this that the semi-colon MUST also be escaped,
     *     as in some properties semi-colon is used for separators.
     *   * Properties using semi-colons: N, ADR, GEO, ORG
     *   * Both ADR and N's individual parts may be broken up further with a
     *     comma.
     *   * Properties using commas: NICKNAME, CATEGORIES
     *
     * vCard 4.0 (rfc6350) says:
     *   * Commas must be escaped.
     *   * Semi-colons may be escaped, an unescaped semi-colon _may_ be a
     *     delimiter, depending on the property.
     *   * Backslashes must be escaped
     *   * Newlines must be escaped as either \N or \n.
     *   * Some compound properties may contain multiple parts themselves, so a
     *     comma within a semi-colon delimited property may also be unescaped
     *     to denote multiple parts _within_ the compound property.
     *   * Text-properties using semi-colons: N, ADR, ORG, CLIENTPIDMAP.
     *   * Text-properties using commas: NICKNAME, RELATED, CATEGORIES, PID.
     *
     * Even though the spec says that commas must always be escaped, the
     * example for GEO in Section 6.5.2 seems to violate this.
     *
     * iCalendar 2.0 (rfc5545) says:
     *   * Commas or semi-colons may be used as delimiters, depending on the
     *     property.
     *   * Commas, semi-colons, backslashes, newline (\N or \n) are always
     *     escaped, unless they are delimiters.
     *   * Colons shall not be escaped.
     *   * Commas can be considered the 'default delimiter' and is described as
     *     the delimiter in cases where the order of the multiple values is
     *     insignificant.
     *   * Semi-colons are described as the delimiter for 'structured values'.
     *     They are specifically used in Semi-colons are used as a delimiter in
     *     REQUEST-STATUS, RRULE, GEO and EXRULE. EXRULE is deprecated however.
     *
     * We unescape backslash-escaped double quotes, t, r, b, f as well, even
     * though those cannot occur in standards-compliant documents, because
     * libical generates those.
     *
     * Now for the parameters
     *
     * If delimiter is not set (null) this method will just return a string.
     * If it's a comma or a semi-colon the string will be split on those
     * characters, and always return an array.
     *
     * @param string $input
     * @param string $delimiter
     * @return string|string[]
     */
    static public function unescapeValue($input, $delimiter = ';') {

        $regex = '#  (?: (\\\\ (?: \\\\ | N | n | ; | , | t | r | b | f | " ) )';
        if ($delimiter) {
            $regex .= ' | (' . $delimiter . ')';
        }
        $regex .= ') #x';

        $matches = preg_split($regex, $input, -1, PREG_SPLIT_DELIM_CAPTURE | PREG_SPLIT_NO_EMPTY);

        $resultArray = array();
        $result = '';

        foreach($matches as $match) {

            switch ($match) {
                case '\\\\' :
                    $result .='\\';
                    break;
                case '\N' :
                case '\n' :
                    $result .="\n";
                    break;
                case '\;' :
                    $result .=';';
                    break;
                case '\,' :
                    $result .=',';
                    break;
                case $delimiter :
                    $resultArray[] = $result;
                    $result = '';
                    break;
                case '\t' :
                    $result .="\t";
                    break;
                case '\r' :
                    $result .="\r";
                    break;
                case '\b' :
                    $result .="\010";
                    break;
                case '\f' :
                    $result .="\f";
                    break;
                case '\"' :
                    $result .='"';
                    break;
                default :
                    $result .= $match;
                    break;

            }

        }

        $resultArray[] = $result;
        return $delimiter ? $resultArray : $result;

    }

    /**
     * Unescapes a parameter value.
     *
     * vCard 2.1:
     *   * Does not mention a mechanism for this. In addition, double quotes
     *     are never used to wrap values.
     *   * This means that parameters can simply not contain colons or
     *     semi-colons.
     *
     * vCard 3.0 (rfc2425, rfc2426):
     *   * Parameters _may_ be surrounded by double quotes.
     *   * If this is not the case, semi-colon, colon and comma may simply not
     *     occur (the comma used for multiple parameter values though).
     *   * If it is surrounded by double-quotes, it may simply not contain
     *     double-quotes.
     *   * This means that a parameter can in no case encode double-quotes, or
     *     newlines.
     *
     * vCard 4.0 (rfc6350)
     *   * Behavior seems to be identical to vCard 3.0
     *
     * iCalendar 2.0 (rfc5545)
     *   * Behavior seems to be identical to vCard 3.0
     *
     * Parameter escaping mechanism (rfc6868) :
     *   * This rfc describes a new way to escape parameter values.
     *   * New-line is encoded as ^n
     *   * ^ is encoded as ^^.
     *   * " is encoded as ^'
     *
     * @param string $input
     * @return void
     */
    private function unescapeParam($input) {

        return
            preg_replace_callback(
                '#(\^(\^|n|\'))#',
                function($matches) {
                    switch($matches[2]) {
                        case 'n' :
                            return "\n";
                        case '^' :
                            return '^';
                        case '\'' :
                            return '"';

                    // @codeCoverageIgnoreStart
                    }
                    // @codeCoverageIgnoreEnd
                },
                $input
            );
    }

    /**
     * Gets the full quoted printable value.
     *
     * We need a special method for this, because newlines have both a meaning
     * in vCards, and in QuotedPrintable.
     *
     * This method does not do any decoding.
     *
     * @return string
     */
    private function extractQuotedPrintableValue() {

        // We need to parse the raw line again to get the start of the value.
        //
        // We are basically looking for the first colon (:), but we need to
        // skip over the parameters first, as they may contain one.
        $regex = '/^
            (?: [^:])+ # Anything but a colon
            (?: "[^"]")* # A parameter in double quotes
            : # start of the value we really care about
            (.*)$
        /xs';

        preg_match($regex, $this->rawLine, $matches);

        $value = $matches[1];
        // Removing the first whitespace character from every line. Kind of
        // like unfolding, but we keep the newline.
        $value = str_replace("\n ", "\n", $value);

        // Microsoft products don't always correctly fold lines, they may be
        // missing a whitespace. So if 'forgiving' is turned on, we will take
        // those as well.
        if ($this->options & self::OPTION_FORGIVING) {
            while(substr($value,-1) === '=') {
                // Reading the line
                $this->readLine();
                // Grabbing the raw form
                $value.="\n" . $this->rawLine;
            }
        }

        return $value;

    }

}
